source/common/upstream/outlier_detection_impl.cc

```diff
@@ -94,16 +110,94 @@ Http::Code DetectorHostMonitorImpl::resultToHttpCode(Result result) {
   case Result::REQUEST_FAILED:
     http_code = Http::Code::InternalServerError;
     break;
-  case Result::SERVER_FAILURE:
-    http_code = Http::Code::ServiceUnavailable;
+  case Result::CONNECT_SUCCESS:
+    NOT_REACHED_GCOVR_EXCL_LINE; // SUCCESS is used to report connection level success, not
+                                 // transaction
     break;
   }
 
   return http_code;
 }
 
-void DetectorHostMonitorImpl::putResult(Result result) {
-  putHttpResponseCode(enumToInt(resultToHttpCode(result)));
+// Method is called by putResult when external and local origin errors
+// are not treated differently. All errors are mapped to HTTP codes.
+void DetectorHostMonitorImpl::putResultNoLocalExternalSplit(Result result,
+                                                            absl::optional<uint64_t> code) {
+  switch (result) {
+  case Result::CONNECT_SUCCESS:
+    // SUCCESS means that connection to host was established, but not transaction still can fail.
+    // HTTP server may return 5xx or DB server may indicate error.
+    break;
+  default:
+    putHttpResponseCode(code ? code.value() : enumToInt(resultToHttpCode(result)));
+    break;
+  }
+}
+
+// Method is called by putResult when external and local origin errors
+// are treated separately. Local origin errors have separate counters and
+// separate success rate monitor.
+void DetectorHostMonitorImpl::putResultWithLocalExternalSplit(Result result,
+                                                              absl::optional<uint64_t>) {
+  switch (result) {
+  // SUCCESS is used to report success for connection level. Server may still respond with
+  // error, but connection to server was OK.
+  case Result::CONNECT_SUCCESS:
+    return localOriginNoFailure();
+  // Connectivity related errors.
+  case Result::TIMEOUT:
+  case Result::CONNECT_FAILED:
+    return localOriginFailure();
+  // REQUEST_FAILED is used when connection to server was successful, but transaction on server
+  // level failed. Since it it similar to HTTP 5xx, map it to 5xx handler.
+  case Result::REQUEST_FAILED:
+    // map it to http code and call http handler.
+    return putHttpResponseCode(500);
```

Comment:
should we be using 503 here? Can we use `Code` enums here and below?

Reply:
Done: converted to enum class Http::Code.